### PR TITLE
[Merged by Bors] - added a dedicated queue per node

### LINKED ISSUE #1577: The time it takes to handle a single message is at least the time it takes to send to the slowest peer
There is a single queue for broadcasting gossip messages. When we dequeue, we send the message to all peers concurrently and wait for the message to be sent to all peers before we pass to the next question in the queue.
This means that we always pay at least the time it takes to send to the slowest peer we have. The more peers we have, the more time it takes to handle a single message.

We should discuss this issue and think about how we should fix it.

### PR DESCRIPTION
**Pull request by antonlerner**
## Motivation
Closes #1577 

## Changes
added a dedicated queue per node

## Test Plan
added test to assert messages coming from queue

### CODE REVIEW

File: p2p/net/conn.go [hunk 1]
```diff
@@ -192,7 +197,34 @@ func (c *FormattedConnection) publish(message []byte) {
 //	return cancel
 //}
 
+func (c *FormattedConnection) sendListener() {
+	for {
+		select {
+		case buf := <-c.messages:
+			//todo: we are hiding the error here...
+			err := c.SendSock(buf)
```
Comment (gavraz) [inline p2p/net/conn.go]: You are making it worse than it was.
Before, we were sending on n sockets concurrently and waiting for the slowest.
Now, you are sending on only one socket concurrently and as long as the buffer is not full, we are not waiting at all.
So we have two issues here:
1. We are not using more than one socket concurrently.
2. When the buffer will be full we will do the same/worse - block until we are done sending enough message from the buffer.

File: p2p/net/conn.go [hunk 2]
```diff
@@ -229,9 +261,13 @@ func (c *FormattedConnection) closeUnlocked() error {
 // Close closes the connection (implements io.Closer). It is go safe.
 func (c *FormattedConnection) Close() error {
 	c.wmtx.Lock()
+	defer c.wmtx.Unlock()
```
Comment (gavraz) [inline p2p/net/conn.go]: Note that "Close" cannot be called more than once safely.

File: p2p/net/conn.go [hunk 3]
```diff
@@ -229,9 +261,13 @@ func (c *FormattedConnection) closeUnlocked() error {
...
 	err := c.closeUnlocked()
```
Comment (gavraz) [inline p2p/net/conn.go]: Does calling closeUnlocked more than once causes an error for sure? if not then you might try closing c.stopSending more than once.

File: p2p/net/conn.go [hunk 4]
```diff
@@ -33,6 +33,8 @@ const (
 	Remote
 )
 
+const MessageQueueSize = 50
```
Comment (gavraz) [inline p2p/net/conn.go:36]: Discuss with yosher what value it should be. Do we want to export it to config?

**Review by gavraz — APPROVED** [review]
Looks good. I suggest Yosher gives another look. @y0sher

File: p2p/net/conn.go (hunk 4 above)
Comment (gavraz) [reply p2p/net/conn.go:36]: @y0sher

File: p2p/net/conn.go [hunk 5]
```diff
@@ -55,21 +57,23 @@ type Connection interface {
...
+	session     NetworkSession
+	deadline    time.Duration
+	timeout     time.Duration
+	r           formattedReader
+	wmtx        sync.Mutex
+	w           formattedWriter
+	closed      bool
+	deadliner   deadliner
+	messages    chan []byte
+	stopSending chan bool
```
Comment (y0sher) [inline p2p/net/conn.go]: if you don't read the value this could just be `chan struct`

File: p2p/net/conn.go [hunk 6]
```diff
@@ -192,7 +197,34 @@ func (c *FormattedConnection) publish(message []byte) {
...
+		case <-c.stopSending:
+			return
+
+		}
+	}
+}
+
 func (c *FormattedConnection) Send(m []byte) error {
+	c.wmtx.Lock()
+	if c.closed {
```
Comment (y0sher) [inline p2p/net/conn.go:218]: this happens again in `SendSock` do we really need that ?

File: p2p/net/conn.go [hunk 7]
```diff
@@ -119,9 +122,11 @@ func newConnection(conn readWriteCloseAddresser, netw networker,
 		deadliner:    conn,
 		networker:    netw,
 		session:      session,
+		messages:     make(chan []byte, MessageQueueSize),
+		stopSending:  make(chan bool),
 		msgSizeLimit: msgSizeLimit,
 	}
-
+	go connection.sendListener()
```
Comment (y0sher) [inline p2p/net/conn.go:129]: maybe put this inside `beginEventProcessing` ? and before stating jut use `SendSock`

File: p2p/net/conn.go (hunk 4 above)
Comment (y0sher) [reply p2p/net/conn.go:36]: I can't calculate an accurate number, let's start with this and test. there are several queues in the way so everything should be buffered.

**Comment by antonlerner** [comment]
bors merge

**Comment by antonlerner** [comment]
bors merge

**Pull request closed without merging**